File: Algorithms/Sorting_Algorithms/Merge_Sort_on_Doubly_Linked_List/Mergesort_doubly.cpp
#include <bits/stdc++.h> 
using namespace std; 
class Node  
{  
    public: 
    int data;  
    Node *next, *prev;  
};  
  
Node *split(Node *head); 
Node *merge(Node *first, Node *second)  
{   
    if (!first)  
        return second;  
    if (!second)  
        return first;  
    if (first->data < second->data)  
    {  
        first->next = merge(first->next,second);  
        first->next->prev = first;  
        first->prev = NULL;  
        return first;  
    }  
    else
    {  
        second->next = merge(first,second->next);  
        second->next->prev = second;  
        second->prev = NULL;  
        return second;  
    }  
}  
   
Node *mergeSort(Node *head)  
{  
    if (!head || !head->next)  
        return head;  
    Node *second = split(head);  
    head = mergeSort(head);  
    second = mergeSort(second);
    return merge(head,second);  
}  
    
void insert(Node **head, int data)  
{  
    Node *temp = new Node(); 
    temp->data = data;  
    temp->next = temp->prev = NULL;  
    if (!(*head))  
        (*head) = temp;  
    else
    {  
        temp->next = *head;  
        (*head)->prev = temp;  
        (*head) = temp;  
    }  
}  
  
void print(Node *head)  
{  
    Node *temp = head;  
    cout<<"Forward Traversal using next poitner\n";  
    while (head)  
    {  
        cout << head->data << " ";  
        temp = head;  
        head = head->next;  
    }  
    cout  << "\nBackward Traversal using prev pointer\n";  
    while (temp)  
    {  
        cout << temp->data << " ";  
        temp = temp->prev;  
    }  
}  
void swap(int *A, int *B)  
{  
    int temp = *A;  
    *A = *B;  
    *B = temp;  
}  
Node *split(Node *head)  
{  
    Node *fast = head,*slow = head;  
    while (fast->next && fast->next->next)  
    {  
        fast = fast->next->next;  
        slow = slow->next;  
    }  
    Node *temp = slow->next;  
    slow->next = NULL;  
    return temp;  
}  
  
int main(void)  
{  
    Node *head = NULL;  
    insert(&head, 5);  
    insert(&head, 20);  
    insert(&head, 4);  
    insert(&head, 3);  
    insert(&head, 30);  
    insert(&head, 10);  
    head = mergeSort(head);  
    cout << "Linked List after sorting\n";  
    print(head);  
    return 0;  
} 

File: Algorithms/Sorting/Merge_Sort_on_Doubly_Linked_List/Mergesort_doubly.py
class Node: 
  def __init__(self,data): 
    self.data=data; 
    self.next=None;
    self.previous=None;

class DoubleLinkedList:
  def __init__(self): 
    self.start=None;
    
  def merge(self,a,b): 
    if (a==None): 
      return b; 
    if (b==None): 
      return a;
    if (a.data < b.data): 
      a.next=self.merge(a.next,b); 
      a.next.previous=a; 
      a.previous=None;
      return a; 
    else: 
      b.next=self.merge(a,b.next); 
      (b.next).previous=b;
      b.previous=None;
      return b;
  def mergesort(self,head): 
    if (head==None): 
      return head; 
    if (head.next==None): 
      return head; 
    b = self.div(head); 
    head = self.mergesort(head); 
    b = self.mergesort(b);
    return self.merge(head,b); 
 
  def div(self,head): 
    first=last=head; 
    while(True): 
      if (first.next==None): 
        break;
      if ((first.next).next==None): 
        break;
      first=(first.next).next;
      last=last.next;
      
    t=last.next;
    last.next=None;
    return t; 
     
  def insert(self,d):  
    n=Node(d); 
    n.next=self.start; 
 
    if (self.start!=None): 
      self.start.previous=n; 
    self.start=n; 
  
  def display(self,node):  
    t=node;
    while(node!=None): 
      print (node.data); 
      t=node; 
      node=node.next;

ob=DoubleLinkedList();
inp=int(input("Enter the number of nodes you want to enter:"));
for i in range(inp):
        j=int(input("Enter the data of the node:"));
        ob.insert(j);
ob.start=ob.mergesort(ob.start) 
print ("\nDisplaying the double linked list after merge sort (in ascending order):");
ob.display(ob.start);